src/util/voxel_io.cpp

```diff
@@ -1,105 +1,240 @@
+    read_int(file, name_len);
+
+    LOG_DEBUG(logger, "Voxel save name length: {}", name_len);
+
+    std::string name(name_len, '\0');
+    file.read(name.data(), name_len);
+
+    LOG_INFO(logger, "Voxel save name: {}", name);
+
+    // Get voxel grid size
+    uint32_t x_max, y_max, z_max;
+
+    read_int(file, x_max); // x
+    read_int(file, z_max); // z
+    read_int(file, y_max); // y
+
+    size = {x_max, y_max, z_max};
+    data.resize(x_max * y_max * z_max);
+
+    LOG_DEBUG(logger, "Voxel grid size: {X} x {Y} x {Z}", x_max, y_max, z_max);
+
+    // Get voxel grid center
+    int32_t x_center, y_center, z_center;
+
+    read_int(file, x_center); // x
+    read_int(file, z_center); // z
+    read_int(file, y_center); // y
+
     center = {x_center, y_center, z_center};
-    std::cout << "    max X: " << X_max << std::endl;
-    std::cout << "    max Y: " << Y_max << std::endl;
-    std::cout << "    max Z: " << Z_max << std::endl;
-    data.resize(X_max * Y_max * Z_max);
-    int tiles_read = 0;
-    if (static_cast<bool>(compression)) {
-        std::cout << "There is compression!" << std::endl;
-    } else {
-        data.resize(size[0] * size[1] * size[2]);
-        for (size_t x = 0; x < size[0]; x++)
-            for (size_t z = 0; z < size[2]; z++)
-                for (size_t y = size[1] - 1; y < size[1]; y--) {
-                    // int32_t* color = (int32_t*) malloc(32);
-                    fread(v, sizeof(uint8_t), 4, file);
-                    // fread(&color, 32, 1, file); // read the color
-                    data[(x * static_cast<int>(size[1]) + y) * size[2] + z] =
-                        compress_color(v);
-                    tiles_read++;
-                }
-    }
-    std::cout << "Voxels read: " << tiles_read << std::endl;
-    // free(void_);
-    // free(compression);
-    // free(name_len);
-    // free(name);
-    delete[] name;
-
-    fclose(file);
-    return 0;
-}
 
-VoxelObject::VoxelObject(const std::string path) {
-    int test = from_qb(path, data_, center_, size_);
-    ok_ = (test == 0);
+    LOG_DEBUG(
+        logger, "Voxel grid center: ({X}, {Y}, {Z})", x_center, y_center, z_center
+    );
+
+    // Read the voxels themselves
+    LOG_DEBUG(logger, "Reading voxels");
+
+    size_t voxels_read = 0;
+    for (size_t x = 0; x < size[0]; x++)
+        for (size_t z = 0; z < size[2]; z++)
+            for (size_t y = size[1] - 1; y < size[1]; y--) {
+                uint32_t raw_color;
+
+                read_int(file, raw_color);
+                data[(x * x_max + y) * y_max + z] = parse_color(raw_color);
+
+                voxels_read++;
+            }
+
+    LOG_INFO(logger, "Voxels read: {}", voxels_read);
 }
 
-VoxelObject::VoxelObject(const std::filesystem::path fs_path) {
-    int test = from_qb(fs_path.string(), data_, center_, size_);
-    ok_ = (test == 0);
+// This is partially from goxel with GPL license
+void
+to_qb(const std::string path, terrain::Terrain ter, bool compression)
```

Comment:
Probably still want this templated. Might generate some entity, and want to save it to qb to test.

Reply:
Well then you need a [concept](https://en.cppreference.com/w/cpp/language/constraints).

Reply:
Either terrain inherits from a virtual class, or we check that three functions, `get_size`, `get_offset`, and `get_voxel`, take the right number of arguments, and return the right types.